pyro/infer/traceenum_elbo.py

```diff
@@ -22,36 +22,61 @@
-                enum_dims.append(len(site["fn"].event_shape) - len(site["value"].shape))
-    if not enum_logprobs:
-        return costs
+                enum_sites.setdefault(ordering[name], []).append(site)
+                enum_dims.append(site["fn"].event_dim - site["value"].dim())
+    if not enum_sites:
+        return OrderedDict((t, [site["log_prob"] for site in sites_t])
+                           for t, sites_t in cost_sites.items())
 
-    # Marginalize out all enumerated variables.
+    # Marginalize out all variables that have been enumerated in the model.
     enum_boundary = max(enum_dims) + 1
-    marginal_costs = OrderedDict((t, []) for t in costs)
+    assert enum_boundary <= 0
+    marginal_costs = OrderedDict((t, []) for t in cost_sites)
     with shared_intermediates():
-        for t, costs_t in costs.items():
-            # TODO refine this coarse dependency ordering using time and tensor shapes.
-            logprobs_t = sum((logprobs_u for u, logprobs_u in enum_logprobs.items() if u <= t), [])
-            if not logprobs_t:
-                marginal_costs[t] = costs_t
+        for t, sites_t in cost_sites.items():
+            # TODO split log_factors into connected components wrt shared tensor dims.
+            log_factors = []
+            scales = set()
+            for site in sites_t:
+                if site["log_prob"].dim() <= -enum_boundary:
+                    # For site do not depend on an enumerated variable, procede as usual.
+                    marginal_costs[t].append(site["log_prob"])
+                else:
+                    # For sites that depend on an enumerated variable, we need to apply
+                    # the mask inside- and the scale outside- of the log expectation.
+                    cost = scale_and_mask(site["unscaled_log_prob"], mask=site["mask"])
+                    log_factors.append(cost)
+                    scales.add(site["scale"])
+            if not log_factors:
                 continue
-
-            ordinal_shape = broadcast_shape(*(x.shape for x in logprobs_t))
-            for cost_t in costs_t:
-                target_shape = broadcast_shape(cost_t.shape, ordinal_shape)
-                target_shape = target_shape[enum_boundary:] if enum_boundary else ()
-                marginal_costs[t].append(logsumproductexp(logprobs_t + [cost_t], target_shape))
+            for u, sites_u in enum_sites.items():
+                # TODO refine this coarse dependency ordering using time and tensor shapes.
+                if u <= t:
+                    for site in sites_u:
+                        logprob = site["unscaled_log_prob"]
+                        log_factors.append(logprob)
+                        scales.add(site["scale"])
+            # This is only correct if all enumerated things share a common subsampling scale.
+            # Note that we use a cheap weak comparison by id rather than tensor value, because
```

Comment:
Where exactly are we comparing tensors by `id`?

Reply:
This is subtle. `scales` is a `set`. `set` objects compare elements by `.__hash__`-then-`is`-then-`.__eq__`. Since PyTorch tensors hash by id, we only ever compare by `.__hash__` and `is`. This all happens at `scales.add(site["scale"])`.

Reply:
Ahh..I see. Thanks for explaining!